include/dlaf/util_matrix.h

```diff
@@ -21,149 +21,189 @@ constexpr double M_PI = 3.141592;
+                              b_k(mat_b.blockSize()), ").");
+
+      DLAF_ASSERT_WITH_ORIGIN(location, (b_k(mat_a.nrTiles()) == b_k(mat_c.nrTiles())),
+                              "The result of the left multiplication between ", mat_a_name, " and ",
+                              mat_b_name, " cannot be stored in ", mat_c_name, " (rows of matrix ",
+                              mat_a_name, ", ", b_k(mat_a.nrTiles()), ", not equal to that of matrix ",
+                              mat_c_name, ", ", b_k(mat_c.nrTiles()), ").");
+      DLAF_ASSERT_WITH_ORIGIN(location, (b_k(mat_a.size()) == b_k(mat_c.size())),
+                              "The result of the left multiplication between ", mat_a_name, " and ",
+                              mat_b_name, " cannot be stored in ", mat_c_name, " (size of matrix ",
+                              mat_a_name, ", ", b_k(mat_a.size()), ", not equal to that of matrix ",
+                              mat_c_name, ", ", b_k(mat_c.size()), ").");
+      DLAF_ASSERT_WITH_ORIGIN(location, (b_k(mat_a.nrTiles()) == b_k(mat_c.nrTiles())),
+                              "The result of the left multiplication between ", mat_a_name, " and ",
+                              mat_b_name, " cannot be stored in ", mat_c_name, " (blocksize of matrix ",
+                              mat_a_name, ", ", b_k(mat_a.blockSize()), ", not equal to that of matrix ",
+                              mat_c_name, ", ", b_k(mat_c.blockSize()), ").");
+
+      DLAF_ASSERT_WITH_ORIGIN(location, (a_k(mat_b.nrTiles()) == a_k(mat_c.nrTiles())),
+                              "The result of the left multiplication between ", mat_a_name, " and ",
+                              mat_b_name, " cannot be stored in ", mat_c_name, " (rows of matrix ",
+                              mat_b_name, ", ", a_k(mat_b.nrTiles()), ", not equal to that of matrix ",
+                              mat_c_name, ", ", a_k(mat_c.nrTiles()), ").");
+      DLAF_ASSERT_WITH_ORIGIN(location, (a_k(mat_b.size()) == a_k(mat_c.size())),
+                              "The result of the left multiplication between ", mat_a_name, " and ",
+                              mat_b_name, " cannot be stored in ", mat_c_name, " (size of matrix ",
+                              mat_b_name, ", ", a_k(mat_b.size()), ", not equal to that of matrix ",
+                              mat_c_name, ", ", a_k(mat_c.size()), ").");
+      DLAF_ASSERT_WITH_ORIGIN(location, (a_k(mat_b.blockSize()) == a_k(mat_c.blockSize())),
+                              "The result of the left multiplication between ", mat_a_name, " and ",
+                              mat_b_name, " cannot be stored in ", mat_c_name, " (blocksize of matrix ",
+                              mat_b_name, ", ", a_k(mat_b.blockSize()), ", not equal to that of matrix ",
+                              mat_c_name, ", ", a_k(mat_c.blockSize()), ").");
+      break;
```

Comment:
I would go for a simplified version, but I don't know what the others think.

It is just an example, feel free to pick something if you think it is a good idea.

@rasolca @teonnik @sanciop 

```cpp
  auto op = [](const auto& size, const blas::Op op) {
    using size2d_t = decltype(size);

    switch (op) {
      case blas::Op::NoTrans:
        return size2d_t{size.rows(), size.cols()};
      case blas::Op::Trans:
      case blas::Op::ConjTrans:
        return size2d_t{size.cols(), size.rows()};
      default:
        return size2d_t{-1, -1};
    }
  };

  bool A_is_transposed = opA != blas::Op::NoTrans;
  GlobalTileSize A_tiles = op(mat_a.nrTiles(), opA);
  GlobalElementSize A_size = op(mat_a.size(), opA);
  TileElementSize A_blocksize = op(mat_a.blockSize(), opA);

  bool B_is_transposed = opB != blas::Op::NoTrans;
  GlobalTileSize B_tiles = op(mat_b.nrTiles(), opB);
  GlobalElementSize B_size = op(mat_b.size(), opB);
  TileElementSize B_blocksize = op(mat_b.blockSize(), opB);

  GlobalTileSize C_tiles{A_tiles.rows(), B_tiles.cols()};
  GlobalElementSize C_size{A_size.rows(), B_size.cols()};
  TileElementSize C_blocksize{A_blocksize.rows(), B_blocksize.cols()};

  DLAF_ASSERT_WITH_ORIGIN(location, (A_tiles.cols() == B_tiles.rows()), "The matrices '", mat_a_name,
                          "' and '", mat_b_name, "' are not multipliable. ", A_tiles, " x ", B_tiles,
                          ". (tranpose: A=", A_is_transposed, " and B=", B_is_transposed, ")");
  DLAF_ASSERT_WITH_ORIGIN(location, (A_size.cols() == B_size.rows()), "The matrices '", mat_a_name,
                          "' and '", mat_b_name, "' are not multipliable. ", A_size, " x ", B_size,
                          ". (tranpose: A=", A_is_transposed, " and B=", B_is_transposed, ")");
  DLAF_ASSERT_WITH_ORIGIN(location, (A_blocksize.cols() == B_blocksize.rows()), "The matrices '",
                          mat_a_name, "' and '", mat_b_name, "' are not multipliable. ", A_blocksize,
                          " x ", B_blocksize, ". (tranpose: A=", A_is_transposed,
                          " and B=", B_is_transposed, ")");

  DLAF_ASSERT_WITH_ORIGIN(location, C_tiles == mat_c.nrTiles(), "The result of the multiplication '",
                          mat_a_name, "' x '", mat_b_name, "' cannot be stored in '", mat_c_name, "'. ",
                          A_tiles, " x ", B_tiles, " = ", C_tiles, " != ", mat_c.nrTiles(),
                          ". (tranpose: A=", A_is_transposed, " and B=", B_is_transposed, ")");
  DLAF_ASSERT_WITH_ORIGIN(location, C_size == mat_c.size(), "The result of the multiplication '",
                          mat_a_name, "' x '", mat_b_name, "' cannot be stored in '", mat_c_name, "'. ",
                          A_size, " x ", B_size, " = ", C_size, " != ", mat_c.size(),
                          ". (tranpose: A=", A_is_transposed, " and B=", B_is_transposed, ")");
  DLAF_ASSERT_WITH_ORIGIN(location, C_blocksize == mat_c.blockSize(),
                          "The result of the multiplication '", mat_a_name, "' x '", mat_b_name,
                          "' cannot be stored in '", mat_c_name, "'. ", A_blocksize, " x ", B_blocksize,
                          " = ", C_blocksize, " != ", mat_c.blockSize(),
                          ". (tranpose: A=", A_is_transposed, " and B=", B_is_transposed, ")");
```